libyul/optimiser/LabelIDDispenser.h

```diff
@@ -0,0 +1,73 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#pragma once
+
+#include <libyul/ASTLabelRegistry.h>
+
+#include <functional>
+#include <set>
+
+namespace solidity::yul
+{
+
+struct Block;
+class Dialect;
+
+/// Can spawn new `LabelID`s which depend on `LabelID`s from a parent label registry. Once generation is completed,
+/// a new `ASTLabelRegistry` can be generated based on the used subset of spawned and original IDs.
+class LabelIDDispenser
+{
+public:
+	using LabelID = ASTLabelRegistry::LabelID;
+	/// A set of reserved labels may be provided, which is excluded when generating new labels. If a reserved label
+	/// already appears in the label registry and is used as-is in the AST, it will be reused despite it being
+	/// provided here.
+	explicit LabelIDDispenser(
```

Comment:
One important thing to mention is that the original labels will be preserved even if they're not valid Yul identifiers.

And for `generateNewLabels()` I'd add that labels are guaranteed to be valid and not reserved IFF the original registry satisfies this condition. And it's guaranteed that at least no additional invalid/reserved names will be introduced.

Reply:
Makes sense, noted these things.